packages/twenty-server/src/modules/workflow/workflow-executor/workflow-actions/iterator/iterator.workflow-action.ts

```diff
@@ -0,0 +1,155 @@
+import { type WorkflowIteratorActionInput } from 'src/modules/workflow/workflow-executor/workflow-actions/iterator/types/workflow-iterator-action-settings.type';
+import { WorkflowIteratorResult } from 'src/modules/workflow/workflow-executor/workflow-actions/iterator/types/workflow-iterator-result.type';
+import { type WorkflowAction } from 'src/modules/workflow/workflow-executor/workflow-actions/types/workflow-action.type';
+import { WorkflowRunWorkspaceService } from 'src/modules/workflow/workflow-runner/workflow-run/workflow-run.workspace-service';
+
+@Injectable()
+export class IteratorWorkflowAction implements WorkflowActionInterface {
+  constructor(
+    private readonly workflowRunWorkspaceService: WorkflowRunWorkspaceService,
+  ) {}
+
+  async execute(input: WorkflowActionInput): Promise<WorkflowActionOutput> {
+    const { currentStepId: iteratorStepId, steps, context, runInfo } = input;
+
+    const step = steps.find((step) => step.id === iteratorStepId);
+
+    if (!step) {
+      throw new WorkflowStepExecutorException(
+        'Step not found',
+        WorkflowStepExecutorExceptionCode.STEP_NOT_FOUND,
+      );
+    }
+
+    if (!isWorkflowIteratorAction(step)) {
+      throw new WorkflowStepExecutorException(
+        'Step is not an iterator action',
+        WorkflowStepExecutorExceptionCode.INVALID_STEP_TYPE,
+      );
+    }
+
+    const iteratorInput = resolveInput(
+      step.settings.input,
+      context,
+    ) as WorkflowIteratorActionInput;
+
+    const { items, initialLoopStepIds } = iteratorInput;
+
+    // testing purpose, will be removed once the UI is implemented
+    const parsedInitialLoopStepIds = isString(initialLoopStepIds)
+      ? JSON.parse(initialLoopStepIds)
+      : initialLoopStepIds;
+
+    const parsedItems = isString(items) ? JSON.parse(items) : items;
+
+    if (!Array.isArray(parsedItems)) {
+      throw new WorkflowStepExecutorException(
+        'Iterator input items must be an array',
+        WorkflowStepExecutorExceptionCode.INTERNAL_ERROR,
+      );
+    }
+
+    if (parsedInitialLoopStepIds.length === 0 || parsedItems.length === 0) {
+      return {
+        result: {
+          itemsProcessed: 0,
+          nextItemToProcess: undefined,
+          hasProcessedAllItems: true,
+        } satisfies WorkflowIteratorResult,
+      };
+    }
+
+    const workflowRun =
+      await this.workflowRunWorkspaceService.getWorkflowRunOrFail({
+        workflowRunId: runInfo.workflowRunId,
+        workspaceId: runInfo.workspaceId,
+      });
+
+    const stepInfos = workflowRun.state.stepInfos;
+    const existingIteratorStepResult = stepInfos[iteratorStepId]
+      ?.result as WorkflowIteratorResult;
+
+    const itemsProcessed = isDefined(existingIteratorStepResult)
+      ? existingIteratorStepResult.itemsProcessed + 1
+      : 0;
+
+    const hasProcessedAllItems = itemsProcessed === parsedItems.length;
```

Comment:
It sounds stupid but I like when we double check that kind of stop criteria:
```suggestion
    const hasProcessedAllItems = itemsProcessed >== parsedItems.length;
```